### PR TITLE
Propagate namespace label references as environment variables used for version detection

### PR DESCRIPTION
**Pull request by mjgrzybek**
# Description
Namespace can be annotated to add additional mapping:
```
annotations:
  mapping.release.dynatrace.com/version: "metadata.annotations['my-version']"
  mapping.release.dynatrace.com/product: "metadata.labels['app.kubernetes.io/name']"
  mapping.release.dynatrace.com/stage: "metadata.namespace"
  mapping.release.dynatrace.com/build-version: "metadata.labels['release.dynatrace.com/stage']"
```
This will configure the mapping used to get the value of the corresponding environment variable via reference in the injected pod.
References are not validated by operator.

Mapping for the different environment variables:
- `DT_RELEASE_VERSION`
  - mapping: `mapping.release.dynatrace.com/version`
  - fallback to default when mapping is missing
- `DT_RELEASE_PRODUCT`
  - mapping: `mapping.release.dynatrace.com/product`
  - fallback to default when mapping is missing
- `DT_RELEASE_STAGE`
  - mapping: `mapping.release.dynatrace.com/stage`
  - no default, no environment variable is missing
- `DT_RELEASE_BUILD_VERSION`
  - mapping: `mapping.release.dynatrace.com/build-version`
  - no default, no environment variable is missing

## How can this be tested?
Apply dynakube CR with `feature.dynatrace.com/label-version-detection: "true"`
Annotate monitored apps namespace with mappings described above, for example:
```
apiVersion: v1                                                                     
kind: Namespace                                                                    
metadata:                                                                          
  annotations:                                                                     
    mapping.release.dynatrace.com/build-version: metadata.annotations['my-build-version']
    mapping.release.dynatrace.com/product: metadata.annotations['my-product']      
    mapping.release.dynatrace.com/stage: metadata.annotations['my-stage']        
    mapping.release.dynatrace.com/version: metadata.annotations['my-version']      
```
Create pod with annotations set:
```
apiVersion: v1                                                 
kind: Pod                                                      
metadata:                                                      
  annotations:                                                 
    data-ingest.dynatrace.com/injected: "true"                 
    dynakube.dynatrace.com/injected: "true"                    
    my-build-version: my custom build-version from deployment-pod-template 
    my-product: ppp
    my-stage: sss
    my-version: bbb
```
Check if pod's env's env variables have values taken from pod's annotations

## Checklist
- [x] Unit tests have been updated/added
- [x] PR is labeled accordingly
- [x] I have read and understood the [contribution guidelines](https://github.com/Dynatrace/dynatrace-operator/blob/master/CONTRIBUTING.md)

### CODE REVIEW

File: src/webhook/mutation/pod_mutator/oneagent_mutation/config.go [hunk 1]
```diff
@@ -9,12 +9,14 @@ var (
...
-	networkZoneEnv       = "DT_NETWORK_ZONE"
-	proxyEnv             = "DT_PROXY"
-	dynatraceMetadataEnv = "DT_DEPLOYMENT_METADATA"
-	releaseVersionEnv    = "DT_RELEASE_VERSION"
-	releaseProductEnv    = "DT_RELEASE_PRODUCT"
+	preloadEnv             = "LD_PRELOAD"
+	networkZoneEnv         = "DT_NETWORK_ZONE"
+	proxyEnv               = "DT_PROXY"
+	dynatraceMetadataEnv   = "DT_DEPLOYMENT_METADATA"
+	releaseVersionEnv      = "DT_RELEASE_VERSION"
```
Comment (meik99) [inline src/webhook/mutation/pod_mutator/oneagent_mutation/config.go:16]: thats the most nitpicky thing, but I would prefer a newline here for better visbility
```suggestion
	dynatraceMetadataEnv   = "DT_DEPLOYMENT_METADATA"
	
	releaseVersionEnv      = "DT_RELEASE_VERSION"
```

File: src/webhook/mutation/pod_mutator/oneagent_mutation/versionlabel.go [hunk 2]
```diff
@@ -18,11 +26,26 @@ func newVersionLabelMapping(namespace corev1.Namespace) VersionLabelMapping {
 }
 
 func getMappingFromNamespace(namespace corev1.Namespace) VersionLabelMapping {
-	// TODO: Implementation => Parse Namespace annotations into correct format
-	return VersionLabelMapping{}
+	mappingKey2versionLabelKey := map[string]string{
```
Comment (meik99) [inline src/webhook/mutation/pod_mutator/oneagent_mutation/versionlabel.go:29]:
```suggestion
	annotationLabelMap := map[string]string{
```

File: src/webhook/mutation/pod_mutator/oneagent_mutation/versionlabel.go [hunk 3]
```diff
@@ -18,11 +26,26 @@ func newVersionLabelMapping(namespace corev1.Namespace) VersionLabelMapping {
...
+		versionMappingAnnotationName: releaseVersionEnv,
+		productMappingAnnotationName: releaseProductEnv,
+		stageMappingAnnotationName:   releaseStageEnv,
+		buildVersionAnnotationName:   releaseBuildVersionEnv,
+	}
+
+	versionLabelMapping := VersionLabelMapping{}
+	for mappingKey, versionLabelKey := range mappingKey2versionLabelKey {
```
Comment (meik99) [inline src/webhook/mutation/pod_mutator/oneagent_mutation/versionlabel.go:37]:
```suggestion
	for annotation, label := range mappingKey2versionLabelKey {
```

File: src/webhook/mutation/pod_mutator/oneagent_mutation/versionlabel.go [hunk 4]
```diff
@@ -18,11 +26,26 @@ func newVersionLabelMapping(namespace corev1.Namespace) VersionLabelMapping {
...
+		if fieldRef, ok := namespace.Annotations[mappingKey]; ok {
```
Comment (meik99) [inline src/webhook/mutation/pod_mutator/oneagent_mutation/versionlabel.go:38]:
```suggestion
		if fieldRef, hasAnnotation := namespace.Annotations[annotation]; hasAnnotation {
```

**Review by meik99 — DISMISSED** [review]

File: src/webhook/mutation/pod_mutator/oneagent_mutation/versionlabel.go (hunk 4 above)
Comment (mjgrzybek) [reply src/webhook/mutation/pod_mutator/oneagent_mutation/versionlabel.go:38]: `ok` is idiomatic. I also think it's unfortunate but prefer to follow general rules.

File: src/webhook/mutation/pod_mutator/oneagent_mutation/versionlabel.go (hunk 3 above)
Comment (mjgrzybek) [reply src/webhook/mutation/pod_mutator/oneagent_mutation/versionlabel.go:37]: Both annotation and label conceptually are key and value pairs.
In this case, we assign `...Key` as a value and I think keeping naming straightforward is better.

File: src/webhook/mutation/pod_mutator/oneagent_mutation/versionlabel.go (hunk 4 above)
Comment (meik99) [reply src/webhook/mutation/pod_mutator/oneagent_mutation/versionlabel.go:38]: if we use golang conventions the line would look like so 
`if f, ok := n.Annotations[a]; ok {`
we already break a lot of conventions because most of them make code unreadable, so if we don't use `ok` for checking if a key is in a map I think thats okay as well

File: src/webhook/mutation/pod_mutator/oneagent_mutation/versionlabel.go (hunk 3 above)
Comment (meik99) [reply src/webhook/mutation/pod_mutator/oneagent_mutation/versionlabel.go]: how about `annotationKey` and `labelKey`?

File: src/webhook/mutation/pod_mutator/oneagent_mutation/versionlabel.go (hunk 4 above)
Comment (mjgrzybek) [reply src/webhook/mutation/pod_mutator/oneagent_mutation/versionlabel.go]: It was already discussed here: https://github.com/Dynatrace/dynatrace-operator/pull/981#discussion_r926447779

File: src/webhook/mutation/pod_mutator/oneagent_mutation/containers_test.go [hunk 5]
```diff
@@ -150,10 +153,134 @@ func TestReinvokeUserContainers(t *testing.T) {
...
+}
+
+func assertDoesntContainMappings(t *testing.T, unexpectedMappingKeys []string, request *dtwebhook.MutationRequest) {
+	for _, key := range unexpectedMappingKeys {
+		idx := slices.IndexFunc(request.Pod.Spec.Containers[0].Env, func(envvar corev1.EnvVar) bool {
+			return envvar.Name == key
+		})
+		assert.Negative(t, idx, key+" found in container's envvars")
+	}
+}
```
Comment (0sewa0) [inline src/webhook/mutation/pod_mutator/oneagent_mutation/containers_test.go:286]:
```suggestion
func assertNotContainsMappings(t *testing.T, unexpectedMappingKeys []string, request *dtwebhook.MutationRequest) {
	for _, env := range request.Pod.Spec.Containers[0].Env {
		assert.NotContains(t, unexpectedMappingKeys, env.Name)
	}
}
```

lets try not to use fancy experimental stuff, especially when there is a simpler solution

File: src/webhook/mutation/pod_mutator/oneagent_mutation/versionlabel.go [hunk 6]
```diff
@@ -18,11 +26,26 @@ func newVersionLabelMapping(namespace corev1.Namespace) VersionLabelMapping {
...
+
+	return versionLabelMapping
 }
 
 func mergeMappingWithDefault(labelMapping VersionLabelMapping) VersionLabelMapping {
-	// TODO: Implementation => Combine the maps
-	return defaultVersionLabelMapping
+	result := VersionLabelMapping{}
+	maps.Copy(result, defaultVersionLabelMapping)
+	maps.Copy(result, labelMapping)
```
Comment (0sewa0) [inline src/webhook/mutation/pod_mutator/oneagent_mutation/versionlabel.go:49]: does it makes sense to introduce the experimental library, which promises 0 stability to avoid minimal code like
```
func (mapping VersionLabelMapping) Copy(source VersionLabelMapping) {
	for key, value := range source {
		mapping[key] = value
	}
}
```

then the function can be:
```suggestion
	result.Copy(defaultVersionLabelMapping)
	result.Copy(labelMapping)
```

we should move to these fancy functions when they are no longer experimental, but in the meanwhile I feel its less of a headache to just not use the experimental stuff

**Review by 0sewa0 — CHANGES_REQUESTED** [review]

File: src/webhook/mutation/pod_mutator/oneagent_mutation/versionlabel.go (hunk 4 above)
Comment (0sewa0) [reply src/webhook/mutation/pod_mutator/oneagent_mutation/versionlabel.go]: having a special name for `ok` can make sense if its used in a special way, 
but this is a one liner, so having that single line be shorter is preferred

TLDR: `ok` is ok here 🙃

File: src/webhook/mutation/pod_mutator/oneagent_mutation/containers_test.go (hunk 5 above)
Comment (mjgrzybek) [reply src/webhook/mutation/pod_mutator/oneagent_mutation/containers_test.go:286]: Awesome, simpler == better!

**Review by 0sewa0 — APPROVED** [review]
